main.go

```diff
@@ -219,3 +237,137 @@ func getCacheFunc() (cache.NewCacheFunc, error) {
+			return nil, fmt.Errorf("error getting APIExport %q: %w", apiExportName, err)
+		}
+	} else {
+		setupLog.Info("api-export-name is empty - listing")
+		exports := &apisv1alpha1.APIExportList{}
+		if err := apiExportClient.List(ctx, exports); err != nil {
+			return nil, fmt.Errorf("error listing APIExports: %w", err)
+		}
+		if len(exports.Items) == 0 {
+			return nil, fmt.Errorf("no APIExport found")
+		}
+		if len(exports.Items) > 1 {
+			return nil, fmt.Errorf("more than one APIExport found")
+		}
+		apiExport = exports.Items[0]
+	}
+
+	if len(apiExport.Status.VirtualWorkspaces) < 1 {
+		return nil, fmt.Errorf("APIExport %q status.virtualWorkspaces is empty", apiExportName)
+	}
+
+	cfg = rest.CopyConfig(cfg)
+	cfg.Host = apiExport.Status.VirtualWorkspaces[0].URL
+
+	return cfg, nil
+}
+
+func kcpAPIsGroupPresent(restConfig *rest.Config) bool {
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(restConfig)
+	if err != nil {
+		setupLog.Error(err, "failed to create discovery client")
+		os.Exit(1)
+	}
+	apiGroupList, err := discoveryClient.ServerGroups()
+	if err != nil {
+		setupLog.Error(err, "failed to get server groups")
+		os.Exit(1)
+	}
+
+	for _, group := range apiGroupList.Groups {
+		if group.Name == apisv1alpha1.SchemeGroupVersion.Group {
+			for _, version := range group.Versions {
+				if version.Version == apisv1alpha1.SchemeGroupVersion.Version {
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+func ensureRequiredAPIGroupsAndResourcesExist(restConfig *rest.Config) {
+	// Do not start the operator until all of the below is available
+	requiredGroupsAndResources := map[string][]string{
+		"tekton.dev": {
+			"pipelineruns",
+			"taskruns",
+		},
+		"appstudio.redhat.com": {
+			"components",
+			"applications",
+		},
+	}
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(restConfig)
+	if err != nil {
+		setupLog.Error(err, "failed to create discovery client")
+		os.Exit(1)
+	}
+
+	if err := wait.PollImmediate(time.Second*5, time.Minute*5, func() (done bool, err error) {
+		apiGroups, apiResources, err := discoveryClient.ServerGroupsAndResources()
+		if err != nil {
+			setupLog.Error(err, "failed to get ServerGroups using discovery client")
+			return false, nil
+		}
+
+	NextGroup:
+		for requiredGroupName, requiredGroupResources := range requiredGroupsAndResources {
+			// Search for the given group in all groups list
+			for _, group := range apiGroups {
+				if group.Name == requiredGroupName {
+					// Required group exists
+					// Check for required resources of the group
+				NextResource:
+					for _, requiredGroupResource := range requiredGroupResources {
+						for _, apiResource := range apiResources {
+							groupName := strings.Split(apiResource.GroupVersion, "/")[0]
+							if groupName == requiredGroupName {
+								for _, apiResourceInGroup := range apiResource.APIResources {
+									if apiResourceInGroup.Name == requiredGroupResource {
+										continue NextResource
```

Comment:
I guess employing labels and continues like this optimizes going through a a lengthy list of groups, or where some groups have a ton of resources, and let's you use ?

Augmenting your map definition to a map of maps could help with tracking status and avoid the jumping around with continue, where you instead remove resources, then groups from `requiredGroupsAndResources`
- populate `requiredGroupsAndResources`as a map[string]map[string]struct{}
- rather than range'ing over `requiredGroupsAndResources`, just range `apiGroups` and `apiResources` returned from `discoveryClient.ServerGroupsAndResources()`
- skip groups whose name when fed into `val, ok := requiredGroupsAndResources[group.Name]` results in `ok` being `false`
- then, with `val` now being a map of `map[string]struct{}`, after you string machinations, if you find a resource who has a key in `val`, then remove it
- when the map[string]struct{} returned from `requiredGroupsAndResources[group.Name]` is at zero length, you have finished that group, so remove the group key from `requiredGroupsAndResources`
- and when `requiredGroupsAndResources` is empty, you have found everything

Reply:
Your approach is good and makes sense (I was thinking about something similar, however I had `bool` instead of `struct{}`), but my point here was to have the clear "interface". It's more friendly to look and at:
```go
"tekton.dev": {
	"pipelineruns",
	"taskruns",
},
"appstudio.redhat.com": {
	"components",
	"applications",
},
```
than have unclear values for each of the resources.
So, I decided to sacrifice a bit code below, but keep the list clear.
Anyway, thank you @gabemontero for the suggestion. If you have strong opinion that we need to go that way, please let me know and I'll rework.

Reply:
No as I tried to convey in my last comment, I don't feel super strongly about it.  If you want to keep it as is, that is fine.  Long term, assuming what you are doing even works in KCP (i.e. @Michkov mentioned in stand up yesterday it was not working in KCP), when I borrow this use of discovery client and use it in jvm build service, I most likely will try my `map[string]map[string]struct{}` approach, and assuming it works, we can compare.  Long term, it probably makes sense to have such a utility in a common repo so we can share.  And then we can let the community decide which approach it likes better ;-)